pkg/recommendation/capacity.go

```diff
@@ -0,0 +1,286 @@
+	if len(deviceTypeNodeBlockDeviceMap) == 0 {
+		return resultMap
+	}
+
+	// Get min max capacity for each type of block devices.
+	for kind, nodeBlockDeviceListMap := range deviceTypeNodeBlockDeviceMap {
+
+		// nodeCapacityDeviceCountMap contains block device capacity and count
+		// for that capacity for each node (block device is present in that node)
+		// After populating nodeBlockDeviceListMap with data this map will look
+		// like this -
+		// Node-1
+		//     - [100GB, 4]
+		//     - [500GB, 6]
+		//     - [1TB, 4]
+		// Node-2
+		//     - [100GB, 1]
+		//     - [500GB, 8]
+		//     - [1TB, 2]
+		// Node-3
+		//     - [500GB, 5]
+		//     - [1TB, 5]
+		// Node-4
+		//     - [100GB, 8]
+		//     - [500GB, 3]
+		nodeCapacityDeviceCountMap := nodeCapacityCount{}
+
+		// Iterating over nodeBlockDeviceListMap map and populating
+		// nodeCapacityDeviceCountMap
+		for nodeName, blockDeviceList := range nodeBlockDeviceListMap {
+
+			// capacityCountMap contains block device capacity and total count
+			// of that capacity in one node.
+			capacityCountMap := nodeCapacityDeviceCountMap.getOrCreate(nodeName)
+
+			for _, blockDevice := range blockDeviceList {
+				capacity, found := blockDevice.Capacity.AsInt64()
+				if !found {
+					// TODO handle this case
+					continue
+				}
+				// get device count for one capacity and increase it by 1
+				count := capacityCountMap.getOrCreate(capacity)
+				capacityCountMap.update(capacity, count+1)
+			}
+
+			// update nodeCapacityDeviceCountMap for each node.
+			nodeCapacityDeviceCountMap.update(nodeName, capacityCountMap)
+		}
+
+		// put min max result for all the types in a map.
+		resultMap[kind] = nodeCapacityDeviceCountMap.getMinMaxCapacity(r.RaidGroupConfig)
+	}
+	return resultMap
+}
+
+// capacityCount is a typed map contains device capacity
+// and count for that capacity
+type capacityCount map[int64]int64
+
+// getOrCreate returns count of a given capacity. If key not
+// found then it adds that key in map and assign with 0
+func (cc capacityCount) getOrCreate(key int64) int64 {
+	value, found := cc[key]
+	if !found {
+		value = 0
+		cc[key] = value
+	}
+	return value
+}
+
+// update takes key and value as argument and update that map
+// with key value pair.
+func (cc capacityCount) update(key, value int64) {
+	cc[key] = value
+}
+
+// getMinMaxCapacityForRaidGroupConfig returns takes raidgroup config as
+// an input and returns min and max capacity in resource.Quantity format
+func (cc capacityCount) getMinMaxCapacityForRaidGroupConfig(
+	raidConfig types.RaidGroupConfig) (min, max resource.Quantity) {
+	max = resource.Quantity{}
+	max = resource.Quantity{}
+
+	// Ideally caller will check raid group config is valid not
+	// If invalid raid config passed then it will not return error.
+	// It only logs the error and return empty struct
+	if err := raidConfig.Validate(); err != nil {
+		// TODO log the error.
```

Comment:
Same check present in caller so this is one kind of repetition of checks.
If you create the request object using `NewCapacityRequest()` then it checks the same and returns an error. Still, we can return error no issue with that. But again caller needs to log the error and proceed to next. 
After refactoring I am facing this issue, I need to add the same check multiple places to avoid extra computation.